src/gmp_zeroize.rs

```diff
@@ -0,0 +1,78 @@
+use gmp_mpfr_sys::gmp::{
+    allocate_function, free_function, get_memory_functions, reallocate_function,
+    set_memory_functions,
+};
+use std::{ffi::c_void, ptr::addr_of_mut, slice, sync::Once};
+use zeroize::Zeroize;
+
+static ONCE: Once = Once::new();
+
+pub fn setup_zeroize() {
+    ONCE.call_once(do_setup_gmp_zeroize);
+}
+
+fn do_setup_gmp_zeroize() {
+    unsafe {
+        get_memory_functions(
+            addr_of_mut!(GMP_ALLOC),
+            addr_of_mut!(GMP_REALLOC),
+            addr_of_mut!(GMP_FREE),
+        );
+
+        // Check that we received the memory functions from GMP.
+        assert!(
```

Comment:
I understand why these asserts are here, but we usually try to avoid having panics in our code. Our devops people will be confused by a panic message on start-up instead of a cleaner error. It also means that callers into a library have no choice on how to handle this failure but to catch the panic.

I suggest we create a new Error variant, like `FailedToInitialize(String)` (where the string is a message about the reason we failed). And this gets returned to the caller.

Reply:
This asserts an invariant of the code, as opposed to a possible error path. That can still be a Result if someone wants to log it differently.

I removed the second assert because it is actually documented: _Rust guarantees to optimize the following types `T` such that [`Option<T>`] has the same size, alignment, and [function call ABI] as `T`… `fn`_.